client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoConnectionDataSource.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.jdbc;
+
+import io.trino.client.uri.PropertyName;
+
+import javax.sql.DataSource;
+
+import java.io.PrintWriter;
+import java.sql.Connection;
+import java.sql.Driver;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.Properties;
+import java.util.logging.Logger;
+
+import static io.trino.client.uri.PropertyName.CATALOG;
+import static io.trino.client.uri.PropertyName.PASSWORD;
+import static io.trino.client.uri.PropertyName.SCHEMA;
+import static io.trino.client.uri.PropertyName.USER;
+
+public class TrinoConnectionDataSource
```

Comment:
This class needs to be thread-safe.

Reply:
I can see that some pooled implementations are thread safe. I've added RW reentrant lock to our implementation.